runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectMetrics.java

```diff
@@ -0,0 +1,228 @@
+import org.apache.beam.sdk.metrics.MetricUpdates.MetricUpdate;
+import org.apache.beam.sdk.metrics.MetricsMap;
+
+/**
+ * Implementation of {@link MetricResults} for the Direct Runner.
+ */
+class DirectMetrics extends MetricResults {
+
+  private interface DirectMetric<UpdateT, ResultT> {
+    void applyPhysical(UpdateT update);
+    void applyLogical(UpdateT update);
+    ResultT extractPhysical();
+    ResultT extractLogical();
+  }
+
+  private static class DirectCounter implements DirectMetric<Long, Long> {
+    private final AtomicLong physicalValue = new AtomicLong();
+    private final AtomicLong logicalValue = new AtomicLong();
+
+    @Override
+    public void applyPhysical(Long update) {
+      physicalValue.addAndGet(update);
+    }
+
+    @Override
+    public Long extractPhysical() {
+      return physicalValue.get();
+    }
+
+    @Override
+    public void applyLogical(Long update) {
+      logicalValue.addAndGet(update);
+    }
+
+    @Override
+    public Long extractLogical() {
+      return logicalValue.get();
+    }
+  }
+
+  private static class DirectDistribution
+      implements DirectMetric<DistributionData, DistributionResult> {
+    private final AtomicReference<DistributionData> physicalValue =
+        new AtomicReference(DistributionData.ZERO);
+    private final AtomicReference<DistributionData> logicalValue =
+        new AtomicReference(DistributionData.ZERO);
+
+    @Override
+    public void applyPhysical(DistributionData update) {
+      DistributionData previous;
+      do {
+        previous = physicalValue.get();
+      } while (!physicalValue.compareAndSet(previous, previous.add(update)));
+    }
+
+    @Override
+    public DistributionResult extractPhysical() {
+      return physicalValue.get().extractResult();
+    }
+
+    @Override
+    public void applyLogical(DistributionData update) {
+      DistributionData previous;
+      do {
+        previous = logicalValue.get();
+      } while (!logicalValue.compareAndSet(previous, previous.add(update)));
+    }
+
+    @Override
+    public DistributionResult extractLogical() {
+      return logicalValue.get().extractResult();
+    }
+  }
+
+  /** The current values of counters in memory. */
+  private MetricsMap<MetricKey, DirectMetric<Long, Long>> counters =
+      new MetricsMap<MetricKey, DirectMetric<Long, Long>>() {
+        @Override
+        protected DirectMetric<Long, Long> createInstance() {
+          return new DirectCounter();
+        }
+      };
+  private MetricsMap<MetricKey, DirectMetric<DistributionData, DistributionResult>> distributions =
+      new MetricsMap<MetricKey, DirectMetric<DistributionData, DistributionResult>>() {
+        @Override
+        protected DirectMetric<DistributionData, DistributionResult> createInstance() {
+          return new DirectDistribution();
+        }
+      };
+
+  @AutoValue
+  abstract static class DirectMetricQueryResults implements MetricQueryResults {
```

Comment:
Maybe -- MetricQueryResults (and many of these other POJO-like classes/interfaces) will likely move towards being backed by some underlying JSON or Proto or whatever-protocol is being used. Specifically, I would normally expect these results objects to be implemented on top of whatever kind of response we've gotten.